Export LLM documentation revamp

jackzhxng opened the pull request:
### Summary
Structure:
- New Getting Started page
- AOT (export)
  - Old getting started page, which was the NanoGPT tutorial, is moved to `export-custom-llm.md`, with the runner sections removed to add to the `run-with-c-plus-plus.md`
  - New `export-llm.md` page for exporting LLMs with `export_llm` API
- Runtime
  - iOS/Android app docs remain, they detail steps to take after the .pte is generated for running on-device
  - Added a C++ runner page for @larryliu0820  to fill out with the new runner APIs
  - Since the QNN Llama tutorial is highly custom, we are going to leave the export section in it as well instead of dividing like we did for the rest of the tutorials

jackzhxng committed:
Update LLM documentation

jackzhxng commented on docs/source/llm/run-with-c-plus-plus.md:
```diff
@@ -0,0 +1,240 @@
+
+Before you begin, make sure you have:
+
+1. A model exported to `.pte` format using the `export_llm` API as described in [Exporting popular LLMs out of the box](export-llm.md) or []
```
[]?

jackzhxng commented on docs/source/llm/run-with-c-plus-plus.md:
```diff
@@ -0,0 +1,240 @@
+Before you begin, make sure you have:
+
+1. A model exported to `.pte` format using the `export_llm` API as described in [Exporting popular LLMs out of the box](export-llm.md) or []
+2. A tokenizer file compatible with your model
```
Where do we find this?

jackzhxng commented on docs/source/llm/run-with-c-plus-plus.md:
```diff
@@ -0,0 +1,240 @@
+
+1. A model exported to `.pte` format using the `export_llm` API as described in [Exporting popular LLMs out of the box](export-llm.md) or []
+2. A tokenizer file compatible with your model
+3. CMake and a C++ compiler installed
```
Maybe be prescriptive and just mention a specific copmiler

jackzhxng commented on docs/source/llm/run-with-c-plus-plus.md:
````diff
@@ -0,0 +1,240 @@
+};
+```
+
+The `resolve_max_new_tokens` method handles the complex logic of determining how many tokens can be generated based on:
````
```suggestion
The `resolve_max_new_tokens` method handles the logic of determining how many tokens can be generated based on:
```

jackzhxng commented on docs/source/llm/run-with-c-plus-plus.md:
```diff
@@ -0,0 +1,240 @@
+
+## Other APIs
+
+1. **Warmup**: For more accurate timing, perform a warmup run before measuring performance:
```
Would what this does, assume it's to cold start load the model?

larryliu0820 replied:
For benchmarking purpose

jackzhxng commented on docs/source/llm/run-with-c-plus-plus.md:
```diff
@@ -0,0 +1,240 @@
+
+The metadata includes several important configuration parameters:
+
+1. **`enable_dynamic_shape`**: Whether the model supports dynamic input shapes
+2. **`max_seq_len`**: Maximum sequence length the model can handle
+3. **`max_context_len`**: Maximum context length for KV cache
+4. **`use_kv_cache`**: Whether the model uses KV cache for efficient generation
+5. **`use_sdpa_with_kv_cache`**: Whether the model uses the custom op [`torch.ops.llama.sdpa_with_kv_cache.default`](https://github.com/pytorch/executorch/blob/release/0.7/extension/llm/custom_ops/op_sdpa.cpp#L611-L614)
```
Are these in the metadata? Do we and why do we need to specify them for the runner?

larryliu0820 replied:
To control the behaviors of the runner. For this one we may not needed anymore, I can remove it.

jackzhxng commented on docs/source/llm/run-with-c-plus-plus.md:
```diff
@@ -0,0 +1,240 @@
+# Running LLMs with C++
```
Overall I would add a section like this https://github.com/pytorch/executorch/blob/main/examples/models/llama/README.md#step-3-run-on-your-computer-to-validate so the user can just use the C++ runner as a binary too

jackzhxng commented on docs/source/llm/export-llm.md:
````diff
@@ -0,0 +1,269 @@
+  --config path/to/config.yaml
+```
+
+### pt2e (Mengwei to flesh this out)
````
@larryliu0820 reminder on this

jackzhxng reviewed (commented):
Thanks for adding the runner doc @larryliu0820

metascroy commented on docs/source/llm/export-custom-llm.md:
````diff
@@ -0,0 +1,424 @@
+from torchao.quantization.pt2e.quantize_pt2e import convert_pt2e, prepare_pt2e
+```
+
+```python
````
You can also link to XNNPACK docs for XNNPACK quantization: https://github.com/pytorch/executorch/blob/main/docs/source/backends-xnnpack.md#quantization

metascroy commented on docs/source/llm/export-custom-llm.md:
```diff
@@ -0,0 +1,424 @@
+large models on consumer electronics. In particular, large language models, such as Llama2, may require quantizing
+model weights to 4 bits or less.
+
+Leveraging quantization requires transforming the model before export. PyTorch provides the pt2e (PyTorch 2 Export)
```
You can link to the quantization doc page here too: https://github.com/pytorch/executorch/blob/main/docs/source/quantization-overview.md

metascroy commented on docs/source/llm/export-llm.md:
```diff
@@ -0,0 +1,269 @@
+- `torchao:8da(\d+)w`: int8 dynamic activation + int[1-8] weights, for example `torchao:8da5w`
+- `torchao:fpa(\d+)w`: int[1-8] weight only, for example `torchao:fpa4w`
+
+To quantize embeddings, specify either `embedding_quantize: <bitwidth>,<groupsize>`, or for low-bit kernels use `embedding_quantize: torchao:<bitwidth>,<groupsize>`. `bitwidth` must be either 2, 4, or 8.
```
For torchao, bitwidth can be 1-8

shoumikhin commented:
Please make a placeholder for ObjC/Swift runtime APIs and i'll fill it out

jackzhxng committed:
Scott comments

mergennachin reviewed (approved):
export-llm is a great page, thank you for working on this

jackzhxng committed:
Mergen comments

jackzhxng commented:
@shoumikhin sure, feel free to open a PR against this!

digantdesai commented on docs/source/llm/export-custom-llm.md:
````diff
@@ -0,0 +1,424 @@
+
+Create a file called export_nanogpt.py with the following contents:
+
+```python
````
so much code (I know you are just moving here) in an `.md` is just waiting to get outdated. Is there an alternative like a notebook or something  (which is also in CI)?

jackzhxng replied:
Yeah, it'd be good to put into a notebook maybe

digantdesai commented on docs/source/llm/run-with-c-plus-plus.md:
```diff
@@ -0,0 +1 @@
+# TODO(mengwei)
```
Nit: put issue # here?

jackzhxng committed:
Anthony comment

Co-authored-by: Anthony Shoumikhin <anthony@shoumikh.in>

larryliu0820 committed:
Add run with c++ (#12569)

### Summary
[PLEASE REMOVE] See [CONTRIBUTING.md's Pull
Requests](https://github.com/pytorch/executorch/blob/main/CONTRIBUTING.md#pull-requests)
for ExecuTorch PR guidelines.

[PLEASE REMOVE] If this PR closes an issue, please add a `Fixes
#<issue-id>` line.

[PLEASE REMOVE] If this PR introduces a fix or feature that should be
the upcoming release notes, please add a "Release notes: <area>" label.
For a list of available release notes labels, check out
[CONTRIBUTING.md's Pull
Requests](https://github.com/pytorch/executorch/blob/main/CONTRIBUTING.md#pull-requests).

### Test plan
[PLEASE REMOVE] How did you test this PR? Please write down any manual
commands you used and note down tests that you have written if
applicable.

larryliu0820 committed:
Update

jackzhxng committed:
Add run on iOS page

shoumikhin committed:
Update run-on-ios.md

shoumikhin committed:
Update index.md

jackzhxng commented:
@pytorchbot cherry-pick --onto release/0.7 -c docs

mergennachin commented:
@jackzhxng what about the main branch?
